File: code/dynamic_programming/src/binomial_coefficient/binomial_coefficient.js
// dynamic programming | binomial coefficient | javascript js
// Part of Cosmos by OpenGenus Foundation

/*Call the function with desired values of n and k.
'ar' is the array name.*/

function binomialCoefficient(n, k) {
  if (k > n) {
    console.log("Not possible.");
    return;
  }

  const ar = [n + 1];
  for (let i = 0; i < n + 1; i++) {
    ar[i] = [n + 1];
  }

  for (i = 0; i < n + 1; i++) {
    for (let j = 0; j < n + 1; j++) {
      if (i == j || j == 0) {
        ar[i][j] = 1;
      } else if (j > i) {
        break;
      } else {
        ar[i][j] = ar[i - 1][j - 1] + ar[i - 1][j];
      }
    }
  }

  console.log(ar[n][k]);
}

File: code/dynamic_programming/src/binomial_coefficient/binomial_coefficient.py
# dynamic programming | binomial coefficient | Python
# Part of Cosmos by OpenGenus Foundation
import numpy as np

# implementation using dynamic programming
def binomialCoefficient(n, k, C=None):
    if k < 0 or n < 0:
        raise ValueError("n,k must not be less than 0")

    if k > n:
        return 0

    k = min(k, n - k)

    if C is None:
        C = np.zeros((n + 1, k + 1))

    if k == 0:
        C[n, k] = 1

    if C[n, k] == 0:
        C[n, k] = binomialCoefficient(n - 1, k, C=C) + binomialCoefficient(
            n - 1, k - 1, C=C
        )

    return C[n, k]


# test
# print(binomialCoefficient(10,3))
# print(binomialCoefficient(10,6))
# print(binomialCoefficient(10,4))
# print(binomialCoefficient(1,1))
# print(binomialCoefficient(0,0))
# print(binomialCoefficient(0,1))
# print(binomialCoefficient(-1,1))
